### ISSUE TITLE
giou

### ISSUE THREAD

**Issue by ZYN-guy** [open-mmlab/mmdetection#1848]
I add some codes in iou_loss.py
### ### ### ### ### ### ### ### ### ### ### 
import torch
import torch.nn as nn

from mmdet.core import bbox_overlaps
from ..registry import LOSSES
from .utils import weighted_loss

@weighted_loss
def iou_loss(pred, target, eps=1e-6):
    """IoU loss.

    Computing the IoU loss between a set of predicted bboxes and target bboxes.
    The loss is calculated as negative log of IoU.

    Args:
        pred (Tensor): Predicted bboxes of format (x1, y1, x2, y2),
            shape (n, 4).
        target (Tensor): Corresponding gt bboxes, shape (n, 4).
        eps (float): Eps to avoid log(0).

    Return:
        Tensor: Loss tensor.
    """
    ious = bbox_overlaps(pred, target, is_aligned=True).clamp(min=eps)
    loss = -ious.log()
    return loss

@weighted_loss
def bounded_iou_loss(pred, target, beta=0.2, eps=1e-3):
    """Improving Object Localization with Fitness NMS and Bounded IoU Loss,
    https://arxiv.org/abs/1711.00164.

    Args:
        pred (tensor): Predicted bboxes.
        target (tensor): Target bboxes.
        beta (float): beta parameter in smoothl1.
        eps (float): eps to avoid NaN.
    """
    pred_ctrx = (pred[:, 0] + pred[:, 2]) * 0.5
    pred_ctry = (pred[:, 1] + pred[:, 3]) * 0.5
    pred_w = pred[:, 2] - pred[:, 0] + 1
    pred_h = pred[:, 3] - pred[:, 1] + 1
    with torch.no_grad():
        target_ctrx = (target[:, 0] + target[:, 2]) * 0.5
        target_ctry = (target[:, 1] + target[:, 3]) * 0.5
        target_w = target[:, 2] - target[:, 0] + 1
        target_h = target[:, 3] - target[:, 1] + 1

    dx = target_ctrx - pred_ctrx
    dy = target_ctry - pred_ctry

    loss_dx = 1 - torch.max(
        (target_w - 2 * dx.abs()) /
        (target_w + 2 * dx.abs() + eps), torch.zeros_like(dx))
    loss_dy = 1 - torch.max(
        (target_h - 2 * dy.abs()) /
        (target_h + 2 * dy.abs() + eps), torch.zeros_like(dy))
    loss_dw = 1 - torch.min(target_w / (pred_w + eps), pred_w /
                            (target_w + eps))
    loss_dh = 1 - torch.min(target_h / (pred_h + eps), pred_h /
                            (target_h + eps))
    loss_comb = torch.stack([loss_dx, loss_dy, loss_dw, loss_dh],
                            dim=-1).view(loss_dx.size(0), -1)

    loss = torch.where(loss_comb < beta, 0.5 * loss_comb * loss_comb / beta,
                       loss_comb - 0.5 * beta)
    return loss

@LOSSES.register_module
class IoULoss(nn.Module):

    def __init__(self, eps=1e-6, reduction='mean', loss_weight=1.0):
        super(IoULoss, self).__init__()
        self.eps = eps
        self.reduction = reduction
        self.loss_weight = loss_weight

    def forward(self,
                pred,
                target,
                weight=None,
                avg_factor=None,
                reduction_override=None,
                **kwargs):
        if weight is not None and not torch.any(weight > 0):
            return (pred * weight).sum()  # 0
        assert reduction_override in (None, 'none', 'mean', 'sum')
        reduction = (
            reduction_override if reduction_override else self.reduction)
        loss = self.loss_weight * iou_loss(
            pred,
            target,
            weight,
            eps=self.eps,
            reduction=reduction,
            avg_factor=avg_factor,
            **kwargs)
        return loss

@LOSSES.register_module
class BoundedIoULoss(nn.Module):

    def __init__(self, beta=0.2, eps=1e-3, reduction='mean', loss_weight=1.0):
        super(BoundedIoULoss, self).__init__()
        self.beta = beta
        self.eps = eps
        self.reduction = reduction
        self.loss_weight = loss_weight

    def forward(self,
                pred,
                target,
                weight=None,
                avg_factor=None,
                reduction_override=None,
                **kwargs):
        if weight is not None and not torch.any(weight > 0):
            return (pred * weight).sum()  # 0
        assert reduction_override in (None, 'none', 'mean', 'sum')
        reduction = (
            reduction_override if reduction_override else self.reduction)
        loss = self.loss_weight * bounded_iou_loss(
            pred,
            target,
            weight,
            beta=self.beta,
            eps=self.eps,
            reduction=reduction,
            avg_factor=avg_factor,
            **kwargs)
        return loss

@weighted_loss
def giou_bbox_overlaps(bboxes1, bboxes2,mode='iou', is_aligned=False):
    '''Calculate generative overlap between two set of bboxes.

    If ``is_aligned`` is ``False``, then calculate the ious between each bbox
    of bboxes1 and bboxes2, otherwise the ious between each aligned pair of
    bboxes1 and bboxes2.

    Args:
        bboxes1 (Tensor): shape (m, 4)
        bboxes2 (Tensor): shape (n, 4), if is_aligned is ``True``, then m and n
            must be equal.

    Returns:
        gious(Tensor): shape (m, n) if is_aligned == False else shape (m, 1)
    '''
    assert mode in ['iou', 'iof']
    rows = bboxes1.size(0)
    cols = bboxes2.size(0)
    if is_aligned:
        assert rows == cols
    if rows * cols == 0:
        return bboxes1.new(rows, 1) if is_aligned else bboxes1.new(rows, cols)

    if is_aligned:
        overlaps_lt = torch.max(bboxes1[:, :2], bboxes2[:, :2])  # [rows, 2]
        overlaps_rb = torch.min(bboxes1[:, 2:], bboxes2[:, 2:])  # [rows, 2]
        overlaps_wh = (overlaps_rb - overlaps_lt + 1).clamp(min=0)  # [rows, 2]

        closure_lt = torch.min(bboxes1[:, :2], bboxes2[:, :2])  # [rows, 2]
        closure_rb = torch.max(bboxes1[:, 2:], bboxes2[:, 2:])  # [rows, 2]
        closure_wh = (closure_rb - closure_lt + 1).clamp(min=0)  # [rows, 2]
        overlap = overlaps_wh[:, 0] * overlaps_wh[:, 1]
        closure = closure_wh[:, 0] * closure_wh[:, 1]
        area1 = (bboxes1[:, 2] - bboxes1[:, 0] + 1) * (
            bboxes1[:, 3] - bboxes1[:, 1] + 1)
        if mode == 'iou':
            area2 = (bboxes2[:, 2] - bboxes2[:, 0] + 1) * (
                bboxes2[:, 3] - bboxes2[:, 1] + 1)
            union = (area1 + area2 - overlap)
            ious = overlap / union
            gious = ious - (closure-union)/(closure + 1e-7)
        else:
            gious = overlap / area1
    else:
        overlaps_lt = torch.max(bboxes1[:, None, :2], bboxes2[:, :2])  # [rows, cols, 2]
        overlaps_rb = torch.min(bboxes1[:, None, 2:], bboxes2[:, 2:])  # [rows, cols, 2]
        overlaps_wh = (overlaps_rb - overlaps_lt + 1).clamp(min=0)  # [rows, cols, 2]

        closure_lt = torch.min(bboxes1[:, None, :2], bboxes2[:, :2])  # [rows, cols, 2]
        closure_rb = torch.max(bboxes1[:, None, 2:], bboxes2[:, 2:])  # # [rows, cols, 2]
        closure_wh = (closure_rb - closure_lt + 1).clamp(min=0)  # [rows, cols, 2]

        overlap = overlaps_wh[:, :, 0] * overlaps_wh[:, :, 1]
        closure = closure_wh[:, :, 0] * closure_wh[:, :, 1]

        area1 = (bboxes1[:, 2] - bboxes1[:, 0] + 1) * (
            bboxes1[:, 3] - bboxes1[:, 1] + 1)
        if mode == 'iou':
            area2 = (bboxes2[:, 2] - bboxes2[:, 0] + 1) * (
                bboxes2[:, 3] - bboxes2[:, 1] + 1)

            union = (area1[:, None] + area2 - overlap)
            ious = overlap / union
            gious = ious - (closure - union) / (closure + 1e-7)
        else:
            gious = overlap / (area1[:, None])
    return gious

@weighted_loss
def giou_loss(pred, target, eps=1e-6):
    """IoU loss.

    Computing the IoU loss between a set of predicted bboxes and target bboxes.
    The loss is calculated as negative log of IoU.

    Args:
        pred (Tensor): Predicted bboxes of format (x1, y1, x2, y2),
            shape (n, 4).
        target (Tensor): Corresponding gt bboxes, shape (n, 4).
        eps (float): Eps to avoid log(0).

    Return:
        Tensor: Loss tensor.
    """
    gious = giou_bbox_overlaps(pred, target, is_aligned=True).clamp(min=eps)
    loss = 1 - gious
    return loss

@LOSSES.register_module
class GIoULoss(nn.Module):

    def __init__(self, eps=1e-6, reduction='mean', loss_weight=1.0):
        super(GIoULoss, self).__init__()
        self.eps = eps
        self.reduction = reduction
        self.loss_weight = loss_weight

    def forward(self,
                pred,
                target,
                weight=None,
                avg_factor=None,
                reduction_override=None,
                **kwargs):
        if weight is not None and not torch.any(weight > 0):
            return (pred * weight).sum()  # 0
        assert reduction_override in (None, 'none', 'mean', 'sum')
        reduction = (
            reduction_override if reduction_override else self.reduction)
        loss = self.loss_weight * giou_loss(
            pred,
            target,
            weight,
            eps=self.eps,
            reduction=reduction,
            avg_factor=avg_factor,
            **kwargs)
        return loss

if __name__ is "__main__":
    bbox1 = torch.Tensor([[0, 0, 100, 100], [0, 0, 100, 100]])
    bbox2 = torch.Tensor([[0, 50, 150, 100], [0, 50, 150, 100]])
    gious_loss = GIoULoss()
    gloss = gious_loss(bbox1, bbox2)

result:
2019-12-21 17:09:49,536 - INFO - Epoch [3][600/5405]	lr: 0.00100, eta: 6:02:04, time: 0.229, data_time: 0.002, memory: 2181, loss_rpn_cls: 100.8393, loss_rpn_bbox: 2.5541, s0.loss_cls: 17.4686, s0.acc: 99.3623, s0.loss_bbox: -59085117.3616, s1.loss_cls: 2.5208, s1.acc: 99.5537, s1.loss_bbox: 0.3496, s2.loss_cls: 1.0944, s2.acc: 99.3154, s2.loss_bbox: 0.2054, loss: -59084993.1768
2019-12-21 17:10:00,715 - INFO - Epoch [3][650/5405]	lr: 0.00100, eta: 6:01:52, time: 0.224, data_time: 0.002, memory: 2182, loss_rpn_cls: 597.1926, loss_rpn_bbox: 22.0874, s0.loss_cls: 185.0882, s0.acc: 99.7412, s0.loss_bbox: -2991783709.0600, s1.loss_cls: 13.3575, s1.acc: 99.7412, s1.loss_bbox: 1.6016, s2.loss_cls: 5.7576, s2.acc: 99.7412, s2.loss_bbox: 1.0368, loss: -2991782891.8900
2019-12-21 17:10:11,883 - INFO - Epoch [3][700/5405]	lr: 0.00100, eta: 6:01:40, time: 0.223, data_time: 0.002, memory: 2182, loss_rpn_cls: 487.9403, loss_rpn_bbox: 15.2581, s0.loss_cls: 217.9797, s0.acc: 99.7188, s0.loss_bbox: -3862890498.5600, s1.loss_cls: 11.7398, s1.acc: 99.7188, s1.loss_bbox: 0.9912, s2.loss_cls: 5.0283, s2.acc: 99.7188, s2.loss_bbox: 0.7665, loss: -3862889765.1200

loss is bad

**Comment by ZYN-guy** [comment]
I use ab function
 gious = ab(ious - (closure - union)) / ab((closure + 1e-7))
It can work.But ap is very low.Could you can give me some suggestions.Thank you!
